### DISCUSSION TITLE
Compute circulating supply in cron

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#777] [Enhancements - Technical]
Filecoin uses the current estimated circulating supply to compute certain penalties and pledges. Unfortunately, this value is computed _by the client_ every epoch which means:

1. Clients need to compute the circulating in the exact same way or the network will fork.
2. The FVM needs to expose this information via a special syscall (`network::total_fil_circ_supply`).

Proposal:

1. Compute this information at the _end_ of cron in the power actor. It's not particularly expensive to compute. I.e., right here: https://github.com/filecoin-project/builtin-actors/blob/86d7c892295f6f05c133ad6910f2d1ce60f116ca/actors/power/src/lib.rs#L231-L240
2. Store the information in the power actor and return it from the `CurrentTotalPower` method. We're already returning pledge collateral here, so this won't matter too much.

This will actually _reduce_ gas fees in most cases as it'll save a syscall. Any method that calls this sycall _already_ calls `CurrentTotalPower` on the power actor.

**Comment by Stebalien** [comment]
Alternative: we could, alternatively, store this information in the reward actor. But given that we're already handling pledge in the power actor, I feel that there's little point.

**Comment by Stebalien** [comment]
It looks like this can be computed purely in terms of:

1. The burnt funds actor's balance.
2. The market + power locked totals.
3. Some hard coded vesting schedules plus the epoch.

If/when we remove market pledge from this calculation (insignificant anyways), we should be able to compute this from information already available. I.e., `CurrentTotalPower`, the burnt balance, and some constants. So we likely don't even need to add anything to cron.

**Comment by aarshkshah1992** [comment]
@jennijuju @Stebalien Can I pick this one up.....please ?

**Reply by Stebalien** [reply]
Yes! However:

1. Make sure that it fits with the Lotus team priorities.
2. Post an early draft so we can see how _viable_ this is. The proposal here is just a sketch and we may run into some roadblock.

**Reply by aarshkshah1992** [reply]
Is this a priority for the FVM team ?

**Reply by jennijuju** [reply]
@aarshkshah1992  - re priority: I think it is reasonable for you to look into how much optimization this could give us.

**Reply by Stebalien** [reply]
Well, this _isn't_ an optimization. It's a priority for the FVM team because it removes a filecoin-specific wart from the FVM, bringing us one step closer to being a general-purpose VM. That's it.

**Reply by maciejwitowski** [reply]
@aarshkshah1992 did you pick it up in the end? (asking, since I'm still tracking it as "FVM item" and maybe I shouldn't anymore)

**Reply by aarshkshah1992** [reply]
Hey @maciejwitowski I still haven't picked it up. Apologies for not updating the issue.

**Comment by arajasek** [comment]
So I'm pretty sure we've considered this before (maybe just in my head?). My biggest opposition to this idea is the introduction of the details of the vesting schedule to the actors (or even the FVM) code. That is very, _very_ specific to Filecoin mainnet, and rips up the idea that these codebases should be "generic system actors". 

Having said that, I'd love to get this out of clients :P

**Reply by anorth** [reply]
If we put the vesting schedule in state, the actor code will remain generic and configurable to other geneses.

**Reply by Stebalien** [reply]
Hm. We could put it in state but then we'd probably need to record it in, e.g., the power actor's state (I think?).

**Reply by Stebalien** [reply]
Really, we have three ways to go:

1. Hard-code the vesting schedule in code.
2. Record the vesting schedule in some actor's state.
3. Record the multisigs that affect the vesting schedule in some actor's state.

I'm happy with any one of these (although I prefer option 2). But, in the second two cases, I _would_ perform the circulating supply calculation in cron (instead of on-demand as described in https://github.com/filecoin-project/FIPs/discussions/777#discussioncomment-7186364).
